ledger-core/v2/virtual/execute/execute_test.go

```diff
@@ -276,3 +254,168 @@ func TestSMExecute_Semi_IncrementPendingCounters(t *testing.T) {
 
 	mc.Finish()
 }
+
+func TestSMExecute_MigrateBeforeLock(t *testing.T) {
+	var (
+		mc  = minimock.NewController(t)
+		ctx = inslogger.TestContext(t)
+
+		prototype   = gen.UniqueReference()
+		caller      = gen.UniqueReference()
+		callee      = gen.UniqueReference()
+		outgoing    = gen.UniqueID()
+		objectRef   = reference.NewSelf(outgoing)
+		sharedState = &object.SharedState{
+			Info: object.Info{
+				KnownRequests:  make(map[reference.Global]struct{}),
+				ReadyToWork:    smsync.NewConditional(1, "ReadyToWork").SyncLink(),
+				MutableExecute: smsync.NewConditional(1, "MutableExecution").SyncLink(),
+			},
+		}
+	)
+
+	slotMachine := slotmachine.NewControlledSlotMachine(ctx, t, true)
+	slotMachine.PrepareMockedMessageSender(mc)
+	slotMachine.PrepareRunner(mc)
+
+	smExecute := SMExecute{
+		Payload: &payload.VCallRequest{
+			Polymorph:    uint32(payload.TypeVCallRequest),
+			CallType:     payload.CTConstructor,
+			CallFlags:    payload.BuildCallRequestFlags(contract.CallTolerable, contract.CallDirty),
+			CallOutgoing: outgoing,
+
+			Caller:              caller,
+			Callee:              callee,
+			CallSiteDeclaration: prototype,
+			CallSiteMethod:      "New",
+		},
+		Meta: &payload.Meta{
+			Sender: caller,
+		},
+	}
+
+	{
+		catalogWrapper := object.NewCatalogMockWrapper(mc)
+		var catalog object.Catalog = catalogWrapper.Mock()
+		slotMachine.AddInterfaceDependency(&catalog)
+
+		sharedStateData := smachine.NewUnboundSharedData(sharedState)
+		smObjectAccessor := object.SharedStateAccessor{SharedDataLink: sharedStateData}
+
+		catalogWrapper.AddObject(objectRef, smObjectAccessor)
+	}
+
+	{
+		pd := pulse.NewFirstPulsarData(10, longbits.Bits256{})
+		pulseSlot := conveyor.NewPresentPulseSlot(nil, pd.AsRange())
+		slotMachine.AddDependency(&pulseSlot)
+	}
+
+	slotMachine.Start()
+	defer slotMachine.Stop()
+
+	smWrapper := slotMachine.AddStateMachine(ctx, &smExecute)
+
+	require.False(t, smExecute.migrationHappened)
+
+	stepToWait := smExecute.stepTakeLock
+	if !slotMachine.StepUntil(smWrapper.WaitStep(stepToWait)) {
+		t.FailNow()
```

Comment:
I suggest to use panic here, since it's not ordinary exit.
We may get here only if slotMachine.Stop is called before.

Reply:
Same everywhere

Reply:
Then I suggest to put it into StepUntil